proposals/0000-no-toplevel-field-selectors.rst

```diff
@@ -0,0 +1,168 @@
+NoToplevelFieldSelectors
+==============
+
+.. proposal-number:: Leave blank. This will be filled in when the proposal is
+                     accepted.
+.. trac-ticket:: Leave blank. This will eventually be filled with the Trac
+                 ticket number which will track the progress of the
+                 implementation of the feature.
+.. implemented:: Leave blank. This will be filled in with the first GHC version which
+                 implements the described feature.
+.. highlight:: haskell
+.. header:: This proposal is `discussed at this pull request <https://github.com/ghc-proposals/ghc-proposals/pull/160>`_.
+.. sectnum::
+.. contents::
+
+Enabling this Language Extension removes the default of Haskell data
+declarations to generate toplevel field accessor functions for records, such
+that the user can define their own via Generic or positional extraction.
+
+Motivation
+------------
+
+There have been proposals to extend the usage of records in Haskell
+(DuplicateRecordFields, OverloadedRecordFields, DeriveGeneric to name a few),
+but none of them address the original issue where record fields steal the
+toplevel function name for each field for selector purposes. This proposal
+intends to open the design space to this issue by removing the generation of
+these toplevel selector functions.
+
+Possible Alternative Use Cases for Record Names
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+After removing the toplevel selector names, the field names could be used as
+bindings for other values.
+
+- A `generic-lens` equivalent for toplevel lenses
+- namespaced accessors
+- namespaced lenses
+- overloaded-labels based accessors / lenses
+- row types
+- https://github.com/ghc-proposals/ghc-proposals/pull/158
+- ...
+
+Proposed Change Specification
+-----------------------------
+
+Record definition no longer defines toplevel selector functions when the
+``NoToplevelFieldSelectors`` extension is enabled. A
+``NoToplevelFieldSelectors`` pragma for records will also be available, such
+that this extension can be enabled on a per-record basis.
+
+Record construction/update syntax and pattern matching will work as before, the
+disambiguation handled by ``DuplicateRecordFields``. The necessary selectors
+will be part of a ``Record(..)`` export, or can also be named one by one. They
+always have to be associated with the data type though, because there is no more
+toplevel selector.
+
+A function for Template Haskell to go from a record field name to a selector for
+that field will be provided, because it's not possible anymore to go from field
+name directly to selector.
+
+A new `TH` function is added which takes two `Name` and returns an `Exp` which
+is equivalent to `a -> b`, where `a` is the record type specified by the second
+argument, and `b` the contents of the field specified by the first argument.
```

Comment:
Can't one simply use `getField` for this, without needing TH? Or it should be possible to write this function using the existing TH API, I think.

Reply:
So mandate the use of `getField` to defer the resolution to the typechecking phase, as mentioned in the other comment? Works for me.